feat(distance): Add full Distance API support

sylvesterdamgaard opened the pull request:
## Summary

Add comprehensive Distance API support to the Geocodio PHP SDK, enabling distance calculations between coordinates with type-safe enums and a new Coordinate DTO class.

## Changes

### New Enums
- `DistanceMode` - Straightline, Driving, Haversine (backward compat alias)
- `DistanceUnits` - Miles, Km
- `DistanceOrderBy` - Distance, Duration
- `DistanceSortOrder` - Asc, Desc

### New Coordinate DTO
- Type-safe `Coordinate` readonly class with lat/lng validation
- Supports optional ID for tracking origins/destinations in responses
- Converts between string format (`lat,lng,id`) and object format for API

### New Distance Methods
- `distance()` - Calculate from single origin to multiple destinations (GET)
- `distanceMatrix()` - Calculate full matrix: multiple origins × destinations (POST)

### Async Distance Matrix Jobs
- `createDistanceMatrixJob()` - Create background processing job
- `distanceMatrixJobStatus()` - Check job status
- `distanceMatrixJobs()` - List all jobs (paginated)
- `getDistanceMatrixJobResults()` - Get parsed JSON results
- `downloadDistanceMatrixJob()` - Download results to file
- `deleteDistanceMatrixJob()` - Delete a job

### Enhanced Geocoding
- `geocode()` now accepts `destinations` parameter for distance calculations
- `reverse()` now accepts `destinations` parameter for distance calculations
- Full filtering support: maxResults, maxDistance, maxDuration, minDistance, minDuration
- Sorting options: orderBy (distance/duration), sortOrder (asc/desc)

### Documentation
- Comprehensive README documentation for all distance features
- Reorganized TOC: Distance section moved after Uploading Lists

## Test Plan

- [x] Unit tests pass (ArchTest, ListTest, TimeoutConfigurationTest)
- [x] Manual testing with local dev API
  - [x] Basic distance with Coordinate objects
  - [x] Distance with driving mode
  - [x] Distance matrix with custom IDs
  - [x] Async distance matrix job creation, status, results, download, delete
  - [x] Geocode with distance destinations
  - [x] Reverse geocode with distance destinations
- [ ] Integration tests (require GEOCODIO_API_KEY)

sylvesterdamgaard committed:
feat(distance): add distance enums and Coordinate DTO

- Add DistanceMode enum (Straightline, Driving, Haversine)
- Add DistanceUnits enum (Miles, Km)
- Add DistanceOrderBy enum (Distance, Duration)
- Add DistanceSortOrder enum (Asc, Desc)
- Add Coordinate readonly class for type-safe coordinate handling
  with lat/lng validation and optional ID support

sylvesterdamgaard committed:
feat(distance): add full distance API support

Add distance calculation methods:
- distance() - single origin to multiple destinations (GET)
- distanceMatrix() - multiple origins × destinations (POST)

Add async distance matrix job methods:
- createDistanceMatrixJob() - create background job
- distanceMatrixJobStatus() - check job status
- distanceMatrixJobs() - list all jobs
- getDistanceMatrixJobResults() - get parsed results
- downloadDistanceMatrixJob() - download to file
- deleteDistanceMatrixJob() - delete job

Add distance support to geocoding:
- geocode() with destinations parameter
- reverse() with destinations parameter

Features:
- Type-safe enums for mode, units, orderBy, sortOrder
- Coordinate DTO with object format for POST requests
- Haversine→straightline mapping for backward compat
- Filtering: maxResults, maxDistance, maxDuration, etc.
- Custom IDs for tracking origins/destinations

sylvesterdamgaard committed:
test(distance): add distance API tests

Add comprehensive test coverage for:
- GET /distance endpoint with various coordinate formats
- POST /distance matrix endpoint
- Nearest mode with filtering options
- Haversine mode backward compatibility
- Custom ID support for coordinates

sylvesterdamgaard committed:
docs: add distance API documentation

Document all distance features:
- Coordinate format with custom IDs
- Distance mode and units enums
- Add distance to geocode/reverse requests
- Single origin to multiple destinations
- Distance matrix (multiple origins × destinations)
- Nearest mode filtering options
- Async distance matrix jobs

Reorganize TOC: move distance after uploading lists

sylvesterdamgaard committed:
fix(test): update expected address in reverse geocoding test

sylvesterdamgaard committed:
fix(distance): validate lat/lng ranges in isCoordinateString

Prevents misidentifying invalid coordinates with range validation:
- Latitude must be between -90 and 90
- Longitude must be between -180 and 180

This catches obvious non-coordinates like '123, 456' while accepting
valid integer coordinates like '38,-77'.

MiniCodeMonkey commented on README.md:
```diff
@@ -324,6 +332,393 @@ The spreadsheet data will always be deleted automatically after 72 hours if it i
+// Available modes
+DistanceMode::Straightline  // Default - great-circle (as the crow flies)
+DistanceMode::Driving       // Road network routing with duration
+DistanceMode::Haversine     // Alias for Straightline (backward compat)
```
To avoid confusion, let's remove Haversine from the README (the constant can still exist in the codebase but with PHPDoc explaining that it's an alias for Straightline)

MiniCodeMonkey commented on README.md:
```diff
@@ -324,6 +332,393 @@ The spreadsheet data will always be deleted automatically after 72 hours if it i
+        '39.9526,-75.1652,store_philly'
+    ],
+    DistanceMode::Driving,
+    DistanceUnits::Miles
```
nit: Would read cleaner with named parameters, something like this
```suggestion
$response = $geocoder->geocode(
    '1600 Pennsylvania Ave NW, Washington DC',
    destinations: [       // destinations with custom IDs
        '38.9072,-77.0369,store_dc',
        '39.2904,-76.6122,store_baltimore',
        '39.9526,-75.1652,store_philly'
    ],
    distanceMode: DistanceMode::Driving,
    distanceUnits: DistanceUnits::Miles
```

MiniCodeMonkey commented on README.md:
```diff
@@ -324,6 +332,393 @@ The spreadsheet data will always be deleted automatically after 72 hours if it i
+// Reverse geocode with distances
+$response = $geocoder->reverse(
+    query: '38.8977,-77.0365',
+    destinations: [
```
Yes!! named parameters 😍

sylvesterdamgaard committed:
refactor(distance): rename DistanceUnits::Km to Kilometers

API uses 'kilometers' not 'km', update enum to match.

sylvesterdamgaard committed:
Revert "refactor(distance): rename DistanceUnits::Km to Kilometers"

This reverts commit 8b5a1135f942bcf68a2d8f9033b5994d4ade202b.

MiniCodeMonkey commented on README.md:
````diff
@@ -324,6 +332,393 @@ The spreadsheet data will always be deleted automatically after 72 hours if it i
+$geocoder->deleteDistanceMatrixJob($job['id']);
+```
+
+> Note: Billing is based on the `elements_billed` value returned in the response. Driving mode has a 2× multiplier compared to straightline mode.
````
There's no `elements_billed` in the response anymore. We can just say something like:

```suggestion
> Note: Distance lookups are billed separately, please refer to https://www.geocod.io/pricing/ for more information.
```

MiniCodeMonkey commented on src/Concerns/SendsRequests.php:
```diff
@@ -83,6 +88,112 @@ protected function formattedReverseQuery($query)
+        $value = $mode instanceof DistanceMode ? $mode->value : $mode;
+
+        // Map haversine → straightline for backward compatibility
+        return $value === 'haversine' ? 'straightline' : $value;
```
I don't think this conversion is necessary. The API supports both haversine and straightline AFAIK

sylvesterdamgaard committed:
refactor(distance): update API endpoint URLs

- distanceMatrix() now uses POST /distance-matrix
- Async job endpoints moved from /distance-matrix to /distance-jobs

sylvesterdamgaard committed:
fix: resolve PHPStan level 5 errors

- Remove invalid @var annotations on constants
- Change Response to ResponseInterface for proper PSR-7 compliance
- Add never return type to handleException() method

sylvesterdamgaard committed:
fix(test): update expected address in batch reverse geocode test

sylvesterdamgaard committed:
docs: address PR review feedback

- Remove Haversine from enum documentation
- Use named parameters in geocode example
- Update billing note to reference pricing page

sylvesterdamgaard committed:
refactor(distance): remove unnecessary haversine to straightline mapping

The API supports both 'haversine' and 'straightline' mode values directly,
so the SDK-side conversion is redundant.

MiniCodeMonkey committed:
Merge branch 'master' into distance

MiniCodeMonkey committed:
fix: Syntax error after merge conflict was resolved

sylvesterdamgaard committed:
fix(test): align distance tests with actual API response structure

- Update default mode expectation from 'driving' to 'straightline'
- Remove 'elements_billed' assertions (not returned by API)
- Remove 'duration_seconds' assertions (only in driving mode)
- Rename haversine tests to straightline (API returns 'straightline')
- Fix matrix tests to expect 'results' array structure instead of
  separate 'origins', 'destinations', 'durations', 'distances' arrays
- Remove custom ID from origin in custom ID test
- Remove max_duration test (only works with driving mode)
- Change sort test from duration to distance

sylvesterdamgaard committed:
fix(test): use driving mode for custom ID test

sylvesterdamgaard committed:
fix(distance): use destinations[] format in query string

The API requires destinations[]=... format, not destinations[0]=...
Build query string manually to ensure correct format.

sylvesterdamgaard committed:
fix(distance): preserve literal commas in destination values

The API requires unencoded commas in coordinate strings with custom IDs.
Decode %2C back to literal commas after urlencode().

sylvesterdamgaard committed:
fix(test): use letter-only IDs in custom ID test

The API has a known issue parsing custom IDs ending with digits.
Changed test IDs from dest1/dest2 to destA/destB.

Also reverts temporary URL encoding workarounds that didn't fix the issue.
